docs/dev/design/redirects.rst

```diff
@@ -0,0 +1,318 @@
+  For example ``/new/path/?_=``.
+
+  Instead of adding the path automatically,
+  users have to add the ``$rest`` placeholder in the target URL.
+  For example:
+
+  - From: ``/old/path/$rest``
+    To: ``/new/path/$rest``
+
+  - From: ``/old/path/$rest``
+    To: ``/new/path/?page=$rest&foo=bar``
+
+- Per-domain redirects.
+  Do users have the need for this?
+  The main problem is that we were applying the redirect
+  to external domains, if we stop doing that, is there the need for this?
+  We can also try to improve how our built-in redirects work
+  (specially our canonical domain redirect).
+
+Improving page redirects
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+- Allow to redirect to external domains.
+  This can be useful to apply a redirect of a well known path
+  in all versions to another domain.
+
+  For example, ``/security/`` to a their security policy page in another domain.
+
+  This new feature isn't strictly needed,
+  but it will be useful to simplify the explanation of the feature
+  (one less restriction to explain).
+
+- Allow a wildcard at the end of the from path.
+  This will allow users to migrate a whole directory to a new path
+  without having to create an exact redirect for each version.
+
+  Similar to exact redirects, users need to add the ``$rest`` placeholder explicitly.
+  This means that that page redirects are the same as exact redirects,
+  with the only difference that they apply to all versions.
+
+Improving Sphinx redirects
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+These redirects are useful, but we should rename them to something more general,
+since they apply to all types of projects, not just Sphinx projects.
+
+Proposed names:
+
+- HTML URL to clean URL redirect (``file.html`` to ``file/``)
+- Clean URL to HTML URL redirect (``file/`` to ``file.html``)
+
+Other ideas to improve redirects
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- Run forced redirects before built-in redirects.
+  We currently run built-in redirects before forced redirects,
+  this is a problem when moving a whole project to a new domain.
+  For example, a forced redirect like ``/$rest``,
+  won't work for the root URL of the project,
+  since ``/`` will first redirect to ``/en/latest/``.
+
+  But shouldn't be a real problem, since users will still need to
+  handle the ``/en/latest/file/`` paths.
+
+- Run redirects on the edge.
+  Cloudflare allow us to create redirects on the edge,
+  but they have some limitations around the number of
+  redirect rules that can be created.
+
+  And they will be useful for forced exact redirects only,
+  since we can't match a redirect based on the response of the origin server.
+
+- Merge prefix redirects with exact redirects.
+  Prefix redirects are the same as exact redirects with a wildcard at the end.
+
+- Placeholders.
+  I haven't seen users requesting this feature.
+  We can consider adding it in the future.
+  Maybe we can expose the current language and version as placeholders.
```

Comment:
I think this is a good idea, and perhaps removes the need for Page redirects, since it's the same as `/$lang/$version/$path` as an exact redirect?

Reply:
This will be also useful for custom URLconf in the future when we support changing the position of them, like: 

```
From: /:lang/:version/:path
To: /:lang/prefix/:version/:path
```

Reply:
I don't think we will be allowing to change the positions of components, we already discussed why this introduces a lot of complexity and doesn't solve user requirements (they only need to add a prefix, which we already suport).